### ISSUE TITLE
Keras and other deep-learning libraries

### ISSUE THREAD

**Issue by tristandeleu** [fchollet/keras#253]
There is an [interesting discussion](http://www.reddit.com/r/MachineLearning/comments/3aenkf/the_theano_tutorial_confused_me_so_i_made_a/csbwnn8) in the Machine Learning subreddit about Keras and other python/theano frameworks like Lasagne, Pylearn2 and Blocks. Here are some extracts:

> @dwf: Aside from not being flexible enough for my needs, still no automated unit tests. Just a bunch of scripts that get run with who knows what frequency. That ain't no way to live.

> @benanne: Lasagne definitely has a different target audience compared to Keras. Whereas Keras abstracts Theano away (so you only need to know about it if you want to extend it), Lasagne is built to be used in conjunction with Theano, because the latter is such a powerful tool, and we want to expose this power to the user. We're aiming for customisability and Lasagne is definitely geared more to people who want to experiment (e.g. researchers), and people who are already using Theano.
 
> @kastnerkyle: Keras is targeting one group of users, and Theano is meant for a very different group. It is basically analogous to scikit-learn -> numpy : Keras -> Theano. If you are writing new papers on ML (as people who heavily use Theano are) you need additional flexibility, at the cost of additional pain. Applying these things once the math and approach have been fleshed out using a sklearn style estimator is perfect.

> @kastnerkyle: Ideally, you could combine this rapid prototyping with an easy transition to "sklearn API" so that people can actually use your model on new data... even if they need to tweak the guts to extend to new code. It is such a huge PITA to reimplement other people's code for sanity checking.

The overall statement here is that on a easy-to-use/high-level to flexible/low-level spectrum, Keras appears to be closer to the former end. That's essentially what it was built for (@fchollet correct me if I'm wrong).

I really like the `Model` abstraction as it makes the overall code cleaner but on the other hand I also really like the flexibility of Lasagne -- and easier access to layers, parameters sharing. I would like to hear what you think about that (whether you use Keras or other libraries).

**Comment by tleeuwenburg** [comment]
Thanks for the update.

Note I have a merge request sitting there for a while which converts the
existing tests into unit tests where possible which are easily runnable
with py.test. I put this up at the first step in putting the tests on a
better footing, showing the structure of how to separate the manual tests
from the unit tests. I would be very happy, if this is accepted, to then
put in more work to enhance the test coverage and help push that aspect of
things forward.

Regards,
-Tennessee

On 20 June 2015 at 19:52, Tristan Deleu <notifications@github.com> wrote:

> There is an interesting discussion
> <http://www.reddit.com/r/MachineLearning/comments/3aenkf/the_theano_tutorial_confused_me_so_i_made_a/csbwnn8>
> in the Machine Learning subreddit about Keras and other python/theano
> frameworks like Lasagne, Pylearn2 and Blocks. Here are some extracts:
>
> @dwf <https://github.com/dwf>: Aside from not being flexible enough for
> my needs, still no automated unit tests. Just a bunch of scripts that get
> run with who knows what frequency. That ain't no way to live.
>
> @benanne <https://github.com/benanne>: Lasagne definitely has a different
> target audience compared to Keras. Whereas Keras abstracts Theano away (so
> you only need to know about it if you want to extend it), Lasagne is built
> to be used in conjunction with Theano, because the latter is such a
> powerful tool, and we want to expose this power to the user. We're aiming
> for customisability and Lasagne is definitely geared more to people who
> want to experiment (e.g. researchers), and people who are already using
> Theano.
>
> @kastnerkyle <https://github.com/kastnerkyle>: Keras is targeting one
> group of users, and Theano is meant for a very different group. It is
> basically analogous to scikit-learn -> numpy : Keras -> Theano. If you are
> writing new papers on ML (as people who heavily use Theano are) you need
> additional flexibility, at the cost of additional pain. Applying these
> things once the math and approach have been fleshed out using a sklearn
> style estimator is perfect.
>
> @kastnerkyle <https://github.com/kastnerkyle>: Ideally, you could combine
> this rapid prototyping with an easy transition to "sklearn API" so that
> people can actually use your model on new data... even if they need to
> tweak the guts to extend to new code. It is such a huge PITA to reimplement
> other people's code for sanity checking.
>
> The overall statement here is that on a easy-to-use/high-level to
> flexible/low-level spectrum, Keras appears to be closer to the former end.
> That's essentially what it was built for (@fchollet
> <https://github.com/fchollet> correct me if I'm wrong).
>
> I really like the Model abstraction as it makes the overall code cleaner
> but on the other hand I also really like the flexibility of Lasagne -- and
> easier access to layers, parameters sharing. I would like to hear what you
> think about that (whether you use Keras or other libraries).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fchollet/keras/issues/253>.
>



-- 
--------------------------------------------------
Tennessee Leeuwenburg
http://myownhat.blogspot.com/
"Don't believe everything you think"

**Comment by fchollet** [comment]
I don't think it's a productive discussion, because every participant has a dog in the fight.  As a result, the facts stated are inaccurate and the arguments feature reality distortion patterns more commonly found in political debates. Or alternatively, in JS framework discussions on Hacker News.

I will just state four things:

- On purpose. The purpose of Keras is not to be "Deep Learning for everyone", it is to be a great research library. A great research library will feature three things: great simplicity, great expressiveness and great extensibility. I believe Keras does a good job at delivering on these three. So far, a number of people I have heard from, and possibly more that I haven't heard from, have adopted Keras for their research.

- On simplicity & power: incidentally, a great research library will also end up being usable by everyone. Because a great research library is simple and well-designed, a characteristic that is in fact *correlated* with being powerful, not the opposite. If Keras is being used successfully in serious research, *and* is being used as a first step into deep learning by many students, it is a sign that Keras is delivering on its goals. That's good enough for me.

- On being the Scikit-learn of Deep Learning: Keras is scikit-like in that it has a simple API, but it is very different from scikit in that it is designed with extensibility as a core goal (which implies a codebase that is small, simple, readable, modular, and hierarchical). As a Keras user you are *supposed* to actively adapt Keras to what you need to do in your research. That's quite a jump from scikit-learn. Of course, if you merely use Keras through its API, then the comparison makes sense. It's just superficial.

- On abstracting Theano vs. working with Theano: in the future Keras is not based on Theano. In the future, Theano has been replaced with a better alternative, and Keras has switched its backend to this better option. In this future, libraries that were "working with Theano" do not exist anymore. Right now Theano is the best option, because it nails two things: 1) it's Python, and 2) it does autodifferentiation. But it also has major flaws. Nervana will soon release something that might or might not be a replacement for Theano. If it isn't, then maybe the next thing to come along will be.

As for unit tests, sure, we need them. We've been talking about them for months, we're making progress, and we're getting there. Lasagne did not have unit tests 3 months in either. Keras is still in alpha.

In the meantime, Keras is still reliable, because it has a user base several times larger than the Theano-based alternatives. Of course, alpha-testing is no proper way to find bugs, they should be caught automatically. But it's definitely a guarantee.

**Comment by simonhughes22** [comment]
I use Keras for my PhD research, although I haven't tried to extend it yet, but if I knew more Theano I don't think that would be hard, as it offers a pretty consistent API for extending things in a modular pattern. One of the main reasons I use it is it has solid support for both CNN's and RNN's, which few other deep learning libraries have and it's pretty easy to pick up and use out of the box, while also has a lot of options for choosing optimizers, regularizers and so on. I am somewhat bias as I haven't used Lasagne much. keras seems the most actively maintained right now, there are commits daily and new things added all the time. I also really like the passage library, but they seem to have stopped updating that, and it's geared solely towards RNN's I think. I do prefer the way they don't require padding for the RNN's and only require the output dimensions to be specified, but given the more restricted scope of the project that's easier to do.
